Add user and group management UI

iammajid opened the pull request:
Adds a complete user and group management interface for administrators. The user management includes list, detail, create and edit views with full CRUD operations. Users can be created, edited and deleted through the admin UI. The user detail page displays associated devices, legacy devices, groups and vault memberships. Group membership can be managed directly from the user detail view by adding or removing groups. The group management includes list and detail views with member management functionality.
<img width="1706" height="947" alt="Bildschirmfoto 2025-12-15 um 07 52 48" src="https://github.com/user-attachments/assets/04483594-49cb-4681-8cdf-ba888705b3f5" />
<img width="1708" height="947" alt="Bildschirmfoto 2025-12-15 um 07 52 33" src="https://github.com/user-attachments/assets/103a7466-0e00-4b47-9cd0-bff93c387486" />
<img width="1706" height="949" alt="Bildschirmfoto 2025-12-15 um 07 52 59" src="https://github.com/user-attachments/assets/6934ac18-0f2a-4e65-9a81-75d9b8229917" />
<img width="1706" height="949" alt="Bildschirmfoto 2025-12-15 um 07 53 10" src="https://github.com/user-attachments/assets/6f853bec-84ad-424f-9de0-ff3d9eb330a7" />

iammajid committed:
WIP: add UserList, GroupList and basic layout

iammajid committed:
WIP: add UserList, GroupList and basic layout

mindmonk committed:
add UserCreateDialog

mindmonk committed:
add UserCreateDialog

mindmonk committed:
add create group dialog

mindmonk committed:
add create group dialog

mindmonk committed:
add GroupEditDialog

mindmonk committed:
add GroupEditDialog

iammajid committed:
Add dialogs to user and group list

iammajid committed:
Add dialogs to user and group list

mindmonk committed:
created UserDetail and GroupDetail

mindmonk committed:
created UserDetail and GroupDetail

iammajid committed:
Adjust GroupDetail layout

iammajid committed:
Adjust GroupDetail layout

mindmonk committed:
last used user or group tab saved in localStorage

mindmonk committed:
last used user or group tab saved in localStorage

mindmonk committed:
add GroupAccessDialog
implement pagination for user list in GroupDetail

mindmonk committed:
add GroupAccessDialog
implement pagination for user list in GroupDetail

mindmonk committed:
implement add/remove group member dialog
group detail improvment

mindmonk committed:
implement add/remove group member dialog
group detail improvment

mindmonk committed:
last added at top sort and trash delete button

mindmonk committed:
last added at top sort and trash delete button

mindmonk committed:
codeclean up and vault section with pagination and search

mindmonk committed:
codeclean up and vault section with pagination and search

iammajid committed:
Add UserEdit

iammajid committed:
Add UserEdit

mindmonk committed:
add GroupEdit

mindmonk committed:
add GroupEdit

iammajid committed:
Add UserEditCreate component with form validation

iammajid committed:
Add UserEditCreate component with form validation

mindmonk committed:
undo some GroupDetail changes

mindmonk committed:
undo some GroupDetail changes

mindmonk committed:
member pagination, pre and next button displayed when needed

mindmonk committed:
member pagination, pre and next button displayed when needed

mindmonk committed:
vaults display below info on desktop and below members on mobile

mindmonk committed:
vaults display below info on desktop and below members on mobile

mindmonk committed:
Merge branch 'develop' into feature/user-and-groups

mindmonk committed:
unify list pagination & improve vault list style

mindmonk committed:
unify list pagination & improve vault list style

mindmonk committed:
improve group info section

mindmonk committed:
improve group info section

mindmonk committed:
extend user details with multiple new sections and optimized styling

mindmonk committed:
extend user details with multiple new sections and optimized styling

mindmonk committed:
member search and list items with name and username

mindmonk committed:
member search and list items with name and username

mindmonk committed:
style finetuning

mindmonk committed:
style finetuning

iammajid committed:
Small cleanup and restructure group/user components

iammajid committed:
Small cleanup and restructure group/user components

mindmonk committed:
hide header of empty lists, fix vaults list nothing found

mindmonk committed:
hide header of empty lists, fix vaults list nothing found

mindmonk committed:
style improvement
new components for group and user detail sections

mindmonk committed:
style improvement
new components for group and user detail sections

mindmonk committed:
new group and user info component and style improvment

mindmonk committed:
new group and user info component and style improvment

iammajid committed:
Add profile picture URL input for users and groups

iammajid committed:
Add profile picture URL input for users and groups

mindmonk committed:
mobile view card style

mindmonk committed:
mobile view card style

SailReal committed:
Merge branch 'develop' into feature/user-and-groups

iammajid committed:
Add responsive mobile cards for UserList and GroupList

iammajid committed:
Add responsive mobile cards for UserList and GroupList

mindmonk committed:
Merge branch 'develop' into feature/user-and-groups

mindmonk committed:
new package components/authority and breadcrumbs

mindmonk committed:
new package components/authority and breadcrumbs

mindmonk committed:
Merge branch 'develop' into feature/user-and-groups

mindmonk committed:
unsavedChanges feature for user and group create

mindmonk committed:
unsavedChanges feature for user and group create

mindmonk committed:
fix navbar on subpages

mindmonk committed:
fix navbar on subpages

iammajid committed:
Add user management with Keycloak integration

iammajid committed:
Add user role management and group membership endpoints

iammajid committed:
Add firstName/lastName to user list and detail views

iammajid committed:
Add group management

iammajid committed:
Fix loading state in GroupList

iammajid committed:
Fix group name update

iammajid committed:
Add firstName/lastName to group members and delete dialogs

iammajid committed:
Hide createdOn field from user detail view

iammajid committed:
Add pagination to UserList and GroupList

iammajid committed:
Return 409 Conflict for duplicate username on user creation

iammajid committed:
Optimize sync and improve 409 conflict handling for users

iammajid committed:
Merge branch 'develop' into feature/user-and-groups

iammajid committed:
Fix Jdenticon icon mismatch between UserList and UserDetail

iammajid committed:
Fix TypeScript type errors in user components

iammajid committed:
Merge branch 'develop' into feature/user-and-groups

iammajid committed:
Fix: Replace rfc4648 with @scure/base for base64 encoding

iammajid committed:
Add user feedback for duplicate group name conflict

iammajid committed:
Add integration tests for Users and Groups resources

iammajid committed:
Merge branch 'develop' into feature/user-and-groups

Resolve conflicts in GroupsResourceIT by combining imports
from both branches.

iammajid committed:
Fix formatting and typo issues

iammajid commented on frontend/src/i18n/en-US.json:
```diff
@@ -31,6 +37,10 @@
   "common.remove": "Remove",
   "common.reset": "Reset",
   "common.retry": "Retry",
+  "common.save":"Save",
```
Fixed in commit 56906b0e475c9f89b58acf77fd5860a1389c26ca.

iammajid commented on frontend/src/i18n/en-US.json:
```diff
@@ -31,6 +37,10 @@
   "common.retry": "Retry",
+  "common.save":"Save",
+  "common.saved": "Saved!",
+  "common.search.placeholder":"Search…",
```
Fixed in commit 56906b0e475c9f89b58acf77fd5860a1389c26ca.

iammajid commented on frontend/src/i18n/en-US.json:
```diff
@@ -31,6 +37,10 @@
+  "common.save":"Save",
+  "common.saved": "Saved!",
+  "common.search.placeholder":"Search…",
+  "common.select.placeholder":"Select…",
```
Fixed in commit 56906b0e475c9f89b58acf77fd5860a1389c26ca.

iammajid commented on frontend/src/components/SearchInputGroup.vue:
```diff
@@ -18,9 +18,12 @@
             <div class="flex items-center">
-              <img :src="item.pictureUrl ?? ''" alt="" class="h-6 w-6 shrink-0 rounded-full" >
+              <img v-if="item.pictureUrl" :src="item.pictureUrl" class="w-6 h-6 shrink-0 rounded-full" alt="group icon" />
+              <div v-else class="w-6 h-6 flex items-center rounded-full rounded-full border border-gray-300 bg-white justify-center overflow-hidde">
```
Fixed in commit 56906b0e475c9f89b58acf77fd5860a1389c26ca.

iammajid committed:
Restrict user/group CRUD endpoints to admin role

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -21,23 +37,221 @@ public class GroupsResource {
+		return getMembersWithNames(groupId);
+	}
+
+	@POST
+	@Path("/{groupId}/members/{userId}")
+	@RolesAllowed("user")
+	@Transactional
+	@Operation(summary = "add a user to a group")
+	@APIResponse(responseCode = "204", description = "user added to group")
+	@APIResponse(responseCode = "404", description = "group or user not found")
+	public Response addMember(@PathParam("groupId") @ValidId String groupId, @PathParam("userId") @ValidId String userId) {
+		keycloakAdminService.addUserToGroup(groupId, userId);
+		return Response.noContent().build();
+	}
+
+	@DELETE
+	@Path("/{groupId}/members/{userId}")
+	@RolesAllowed("user")
+	@Transactional
+	@Operation(summary = "remove a user from a group")
+	@APIResponse(responseCode = "204", description = "user removed from group")
+	@APIResponse(responseCode = "404", description = "group or user not found")
+	public Response removeMember(@PathParam("groupId") @ValidId String groupId, @PathParam("userId") @ValidId String userId) {
+		keycloakAdminService.removeUserFromGroup(groupId, userId);
+		return Response.noContent().build();
+	}
```
Fixed in commit a0dc2c5415f8ff8c015bde35b71ca90f5e65a773.

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -21,23 +37,221 @@ public class GroupsResource {
+	@Path("/")
+	@RolesAllowed("user")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	@Transactional
+	@Operation(summary = "create a new group in Keycloak")
+	@APIResponse(responseCode = "201", description = "group created")
+	@APIResponse(responseCode = "400", description = "invalid input")
+	@APIResponse(responseCode = "409", description = "group name already exists")
+	public Response createGroup(@Valid @NotNull CreateGroupDto dto) {
+		try {
+			var groupRepresentation = keycloakAdminService.createGroup(dto.name(), dto.pictureUrl());
+
+			Group group = groupRepo.findById(groupRepresentation.getId());
+			if (group == null) {
+				throw new RuntimeException("Group was created in Keycloak but not found in database after sync");
+			}
+
+			return Response.created(URI.create("./" + group.getId()))
+					.entity(GroupDto.fromEntity(group))
+					.build();
+		} catch (ClientErrorException e) {
+			// Return 409 with specific error message (GROUP_NAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
+					.type(MediaType.TEXT_PLAIN)
+					.build();
+		}
+	}
```
Fixed in commit a0dc2c5415f8ff8c015bde35b71ca90f5e65a773.

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -21,23 +37,221 @@ public class GroupsResource {
+		return GroupDtoWithDetails.from(group, pictureUrl, members, vaults);
 	}
 
+	@PUT
+	@Path("/{groupId}")
+	@RolesAllowed("user")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	@Transactional
+	@Operation(summary = "update a group in Keycloak")
+	@APIResponse(responseCode = "200", description = "group updated")
+	@APIResponse(responseCode = "404", description = "group not found")
+	public GroupDto updateGroup(@PathParam("groupId") @ValidId String groupId, @Valid @NotNull UpdateGroupDto dto) {
+		keycloakAdminService.updateGroup(groupId, dto.name(), dto.pictureUrl());
+
+		Group group = groupRepo.findById(groupId);
+		if (group == null) {
+			throw new jakarta.ws.rs.NotFoundException("Group not found after update: " + groupId);
+		}
+
+		return GroupDto.fromEntity(group);
+	}
+
+	@DELETE
+	@Path("/{groupId}")
+	@RolesAllowed("user")
+	@Transactional
+	@Operation(summary = "delete a group from Keycloak")
+	@APIResponse(responseCode = "204", description = "group deleted")
+	@APIResponse(responseCode = "404", description = "group not found")
+	public Response deleteGroup(@PathParam("groupId") @ValidId String groupId) {
+		keycloakAdminService.deleteGroup(groupId);
+		return Response.noContent().build();
+	}
```
Fixed in commit a0dc2c5415f8ff8c015bde35b71ca90f5e65a773.

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +325,252 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
 		}
 	}
+
+	@POST
+	@Path("/")
+	@RolesAllowed("user")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	@Transactional
+	@Operation(summary = "create a new user in Keycloak")
+	@APIResponse(responseCode = "201", description = "user created")
+	@APIResponse(responseCode = "400", description = "invalid input")
+	@APIResponse(responseCode = "409", description = "user already exists")
+	public Response createUser(@Valid @NotNull CreateUserDto dto) {
```
Fixed in commit a0dc2c5415f8ff8c015bde35b71ca90f5e65a773.

iammajid committed:
Add debug logging for Keycloak connectivity issues

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -21,23 +37,221 @@ public class GroupsResource {
+			}
+		} catch (Exception e) {
+			// continue without keycloak data
+		}
```
Fixed in commit 8e3d5c5c2a61ecbed44da7eb4f6d1ac9a71df48f.

iammajid committed:
Fix attribute preservation and add missing @Transactional

iammajid committed:
Add error display and prevent double-clicks

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,261 @@
+		try {
+			var groupRepresentation = keycloakAdminService.createGroup(dto.name(), dto.pictureUrl());
+
+			Group group = groupRepo.findById(groupRepresentation.getId());
+			if (group == null) {
+				throw new RuntimeException("Group was created in Keycloak but not found in database after sync");
+			}
+
+			return Response.created(URI.create("./" + group.getId()))
+					.entity(GroupDto.fromEntity(group))
+					.build();
+		} catch (ClientErrorException e) {
+			// Return 409 with specific error message (GROUP_NAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
+					.type(MediaType.TEXT_PLAIN)
+					.build();
+		}
+	}
```
Keeping RuntimeException for now.

iammajid commented on frontend/src/components/authority/GroupInfo.vue:
```diff
@@ -0,0 +1,22 @@
+  <section class="bg-white rounded-lg shadow-sm overflow-hidden">
+    <div class="px-6 py-6">
+      <div class="flex flex-col items-center justify-center h-full text-center">
+        <img :src="group.picture" class="w-48 h-48 rounded-full object-cover border border-gray-300 mb-4" />
```
Not needed, AuthorityService.fillInMissingPicture() already generates a Jdenticon fallback.

iammajid committed:
Fix array reference bugs in UserEditCreate role comparison

iammajid committed:
Fix loading state in UserList to distinguish from empty data

iammajid commented on frontend/src/components/SearchInputGroup.vue:
```diff
@@ -8,7 +8,7 @@
             <img v-else :src="selectedItem.pictureUrl ?? ''" alt="" class="w-5 h-5 rounded-full" >
```
Won't fix. The root cause was a field name mismatch, userPicture was used instead of pictureUrl. Fixed in UserAddGroupDialog by using pictureUrl consistently. Jdenticon now works correctly.

iammajid committed:
Fix Jdenticon display in UserAddGroupDialog and add color variation for groups

iammajid committed:
Fix typo in i18n key

iammajid committed:
Restrict user and group management to admin role

iammajid committed:
Sort UserList to show current user at top

iammajid committed:
Optimize vault loading in getUser to fetch only relevant vaults

iammajid committed:
Extract VaultDtoWithRole to VaultResource to avoid code duplication

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +325,252 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+					role
+			);
+		}
+	}
```
Fixed in commit f75bf6250f2f5f394694dd4a55d3644a70dfb9f0.

iammajid commented on frontend/src/i18n/en-US.json:
```diff
@@ -172,6 +190,54 @@
+  "groupEditCreate.roles" : "Roles",
+  "groupEditCreate.validation.required": "This field is required",
+  "groupEditCreate.roles.remove": "Remove roles",
+  "groupEditCreate.roles.chose": "Choose roles",
```
Fixed in commit e2cc27c74dc6eaf28a4e682f1de6b11244ccc60c.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupDto.java:
```diff
@@ -8,17 +8,25 @@ public final class GroupDto extends AuthorityDto {
 
 	public static GroupDto fromEntity(Group group) {
-		return fromEntity(group, false);
+		return new GroupDto(group.getId(), group.getName(), group.getPictureUrl(), null, null);
```
Why do we call here now the constructor directly?

iammajid replied:
That was unintentional. Fixed, now delegating to fromEntity(group, false) again for consistency.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
 	@GET
 	@Path("/")
-	@RolesAllowed("user")
+	@RolesAllowed("admin")
```
Why this change?

iammajid commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
+			// Return 409 with specific error message (GROUP_NAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
+					.type(MediaType.TEXT_PLAIN)
+					.build();
+		}
```
These aren't raw exception details, but predefined error codes.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
+		} catch (ClientErrorException e) {
+			// Return 409 with specific error message (GROUP_NAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
```
Just return 409, do not expose the exception message (can contain not expected content).

iammajid replied:
Fixed, now just returning 409 without exposing the exception message.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
+				if (pictureAttr != null && !pictureAttr.isEmpty()) {
+					pictureUrl = pictureAttr.get(0);
+				}
+			}
```
Why do we read here the picture url directly from keycloak and not from the database?

iammajid replied:
Removed the unnecessary Keycloak call.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
+					} catch (Exception e) {
+						LOG.fine("Could not fetch Keycloak user data for " + user.getId());
+					}
+					return UserDto.UserDtoWithName.from(user, firstName, lastName);
```
Why do we read the names from keycloak and not from the database?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -227,11 +240,32 @@ public Response resetMe() {
+						lastName = keycloakUser.getLastName();
+					} catch (Exception e) {
+						LOG.fine("Could not fetch Keycloak user data for " + user.getId());
+					}
```
Never catch exception, catch the expected one and handle it but do not catch everything.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -227,11 +240,32 @@ public Response resetMe() {
+					try {
+						var keycloakUser = keycloakAdminService.getUser(user.getId());
+						firstName = keycloakUser.getFirstName();
+						lastName = keycloakUser.getLastName();
```
Why do we not use the names from the database but get it from keycloak (for each user)?

iammajid replied:
`firstName` and `lastName` are not stored in our database, only in Keycloak. 

We could either:
1. Add these fields to the User entity and sync them in syncUser()
2. Keep fetching from Keycloak but use batch fetch

SailReal replied:
Let's synchronize them, otherwise we'll create a huge load on Keycloak and cause significant delays in the system. Always calculate roughly one second of turnaround time per request on average (and we are here in a for-each loop, but also without the loop we should synchronize them) VS a few (required) properties in the database.

overheadhunter replied:
I'll take care of this

overheadhunter replied:
fixed in d286558

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+	public Response createUser(@Valid @NotNull CreateUserDto dto) {
+		try {
+			var userRepresentation = keycloakAdminService.createUser(
+					dto.username(),
```
Sometimes we call this property `name` (like in `UserDtoWithName`) and sometimes `username`, I prefer `username` but everywhere.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+		} catch (ClientErrorException e) {
+			// Return 409 with specific error message (EMAIL_EXISTS or USERNAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
```
Do not publish the exception message.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+			var keycloakUser = keycloakAdminService.getUser(userId);
+			createdTimestamp = keycloakUser.getCreatedTimestamp();
+			firstName = keycloakUser.getFirstName();
+			lastName = keycloakUser.getLastName();
```
Why do we read this from keycloak and not from the database?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+		// Fetch legacy devices
+		Set<DeviceResource.DeviceDto> legacyDevices = user.legacyDevices.stream()
+				.map(DeviceResource.DeviceDto::fromEntity)
+				.collect(Collectors.toSet());
```
Could we not write a join to get all those values in less effort?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+				.collect(Collectors.toSet());
+
+		// Fetch roles
+		Set<String> roles = keycloakAdminService.getUserRoles(userId);
```
I think we should persist roles in Hub as well to not have to query them in Keycloak every time when a user gets requested.

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/GroupsResource.java:
```diff
@@ -1,43 +1,242 @@
+					try {
+						var keycloakUser = keycloakAdminService.getUser(user.getId());
+						firstName = keycloakUser.getFirstName();
+						lastName = keycloakUser.getLastName();
```
Why do we read this from keycloak and not from the database?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+	@Transactional
+	@Operation(summary = "update a user in Keycloak")
+	@APIResponse(responseCode = "200", description = "user updated")
+	@APIResponse(responseCode = "403", description = "user has federated identity and cannot be modified")
```
Could also be a permission thing or?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+
+			if (dto.roles() != null) {
+				keycloakAdminService.updateUserRoles(userId, dto.roles());
+			}
```
Shouldn't we merge `updateUser()` and `updateUserRoles()` to save a few requests against Keycloak?

SailReal commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +324,231 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+			return Response.noContent().build();
+		} catch (ForbiddenException | NotFoundException e) {
+			throw e;
+		} catch (Exception e) {
```
exception

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		}
+
+		String locationHeader = response.getHeaderString("Location");
+		String userId = locationHeader.substring(locationHeader.lastIndexOf('/') + 1);
```
Instead of parsing the ID like this, can we not set it initially via `user.setId()`?

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+			for (String groupId : groupIds) {
+				try {
+					userResource.joinGroup(groupId);
+				} catch (Exception e) {
```
exception

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+
+		syncUser(userId);
+
+		return realm.users().get(userId).toRepresentation();
```
Couldn't we return the user from the `syncUser()` to avoid another query against keycloak here?

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+			return !federatedIdentities.isEmpty();
+		} catch (Exception e) {
+			LOG.warn("Failed to check federated identity for user {}", userId, e);
+			return false;
```
Should not return false if the check fails.

Also exception

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+			CredentialRepresentation credential = new CredentialRepresentation();
+			credential.setType(CredentialRepresentation.PASSWORD);
+			credential.setValue(password);
+			credential.setTemporary(false);
```
We should set temporary here to true.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		User user = userRepo.findById(userId);
+		if (user != null) {
+			userRepo.delete(user);
+		}
```
Why do we delete here the user in the database directly? This should be the job of the syncer.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		}
+	}
+
+	public boolean isUserReadOnly(String userId) {
```
Would rename this to something like `federatedUser()` or something like this.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+			throw new RuntimeException("Failed to sync user: " + userId, e);
+		}
+	}
+
+	@Transactional
+	public Group syncGroup(String groupId) {
+		try {
+			RealmResource realm = keycloak.realm(keycloakRealm);
+			GroupResource groupResource = realm.groups().group(groupId);
+			GroupRepresentation keycloakGroup = groupResource.toRepresentation();
+
+			Group dbGroup = groupRepo.findById(groupId);
+			if (dbGroup == null) {
+				dbGroup = new Group();
+				dbGroup.setId(keycloakGroup.getId());
+			}
+
+			dbGroup.setName(keycloakGroup.getName());
+
+			// Sync members
+			var keycloakMembers = groupResource.members();
+			Set<org.cryptomator.hub.entities.Authority> members = new java.util.HashSet<>();
+			for (var member : keycloakMembers) {
+				User dbUser = userRepo.findById(member.getId());
+				if (dbUser != null) {
+					members.add(dbUser);
+				}
+			}
+			dbGroup.setMembers(members);
+
+			groupRepo.persist(dbGroup);
+			groupRepo.flush();
+			return dbGroup;
+		} catch (jakarta.ws.rs.NotFoundException e) {
+			throw new NotFoundException("Group not found in Keycloak: " + groupId);
+		} catch (Exception e) {
+			LOG.error("Failed to sync group {}", groupId, e);
+			throw new RuntimeException("Failed to sync group: " + groupId, e);
+		}
+	}
```
Could we not use for this code the existing syncer code (with some modifciations)? I think this would avoid a lot of duplication.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+	@Transactional
+	public void addUserToGroup(String groupId, String userId) {
+		RealmResource realm = keycloak.realm(keycloakRealm);
+
+		try {
+			realm.groups().group(groupId).toRepresentation();
+		} catch (jakarta.ws.rs.NotFoundException e) {
+			throw new NotFoundException("Group not found: " + groupId);
+		}
```
Is this required? What happens when `userResource.joinGroup(groupId);` is called for an non existent group?

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+			realm.groups().group(groupId).toRepresentation();
+		} catch (jakarta.ws.rs.NotFoundException e) {
+			throw new NotFoundException("Group not found: " + groupId);
+		}
```
Is this required? What happens when `userResource.leaveGroup(groupId);` is called for an non existent group?

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+				if (!rolesToAdd.isEmpty()) {
+					roleMappings.add(rolesToAdd);
+				}
+			}
```
Why do we remove the roles first and add the required ones again? Can we not just add/remove the required ones?

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		}
+
+		String locationHeader = response.getHeaderString("Location");
+		String groupId = locationHeader.substring(locationHeader.lastIndexOf('/') + 1);
```
Would also set here the groupId initially to not parse it from the URL (this behavior could change at any time).

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		String groupId = locationHeader.substring(locationHeader.lastIndexOf('/') + 1);
+
+		syncGroup(groupId);
+
+		return realm.groups().group(groupId).toRepresentation();
```
Would here also return the group via `syncGroup()` to not need the extra request.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+
+		if (name != null && !name.isBlank()) {
+			group.setName(name);
+			groupResource.update(group);
```
Just call `groupResource.update` once, and not again when we set the pictureUrl

overheadhunter replied:
fixed in a987ace

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+
+		if (pictureUrl != null) {
+			try {
+				GroupRepresentation groupForPicture = groupResource.toRepresentation();
```
Why do we call here `groupResource.toRepresentation()` again and do not use the `group` variable? (this can make a request against keycloak)

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		}
+
+		syncGroup(groupId);
+
+		return groupResource.toRepresentation();
```
also here, would return `syncGroup()` to avoid the request against keycloak.

SailReal commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		if (group != null) {
+			groupRepo.delete(group);
+		}
+	}
```
We should not modify the db directly but let the syncer do its job (execute sync for this group).

SailReal reviewed (changes requested):
It looks very promising! :+1:

I skimmed through the first part of the PR (I'll take a closer look at a few sections of this part) and will look at another part tomorrow, but here's some first feedback from the first part for now.

overheadhunter commented on frontend/src/common/formvalidator.ts:
```diff
@@ -0,0 +1,130 @@
+   * Validates email format
+   */
+  static isValidEmail(email: string): boolean {
+    return /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/.test(email.trim());
```
see https://devina.io/redos-checker

```suggestion
    return /^[^\s@]+@[^\s@]+\.[a-z]{2,}$/i.test(email.trim());
```

overheadhunter replied:
fixed in eb31978

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,444 @@
+		credential.setTemporary(false);
+		user.setCredentials(List.of(credential));
+
+		var response = realm.users().create(user);
```
use [try-with-resource](https://www.baeldung.com/java-try-with-resources#using)

overheadhunter replied:
fixed in a987ace

overheadhunter reviewed (changes requested):
just sharing sonarcloud findings, will do an in-depth review tomorrow.

iammajid committed:
Fix GroupDto delegation, 409 response, and use DB for pictureUrl

overheadhunter committed:
Merge branch 'develop' into feature/user-and-groups

# Conflicts:
#	backend/src/test/java/org/cryptomator/hub/api/UsersResourceIT.java

overheadhunter committed:
sync firstname, lastname from kc to db

overheadhunter committed:
cleanup

overheadhunter committed:
more efficient regex (ReDoS safe)

overheadhunter committed:
case-insensitive matching

[ci skip]

overheadhunter committed:
consolidate DTO types, remove leftovers

iammajid committed:
Add cursor-pointer to clickable user/group names

iammajid committed:
Merge branch 'feature/user-and-groups' of github.com:cryptomator/hub into feature/user-and-groups

overheadhunter committed:
consolidate types

iammajid committed:
Merge branch 'feature/user-and-groups' of github.com:cryptomator/hub into feature/user-and-groups

overheadhunter committed:
sync group pictures

iammajid committed:
Merge branch 'feature/user-and-groups' of github.com:cryptomator/hub into feature/user-and-groups

overheadhunter committed:
add `picture` user attribute to realm.json

iammajid committed:
Merge branch 'feature/user-and-groups' of github.com:cryptomator/hub into feature/user-and-groups

iammajid committed:
Fix SonarCloud accessibility issues

overheadhunter committed:
cleanup

overheadhunter committed:
Merge branch 'feature/user-and-groups' of https://github.com/cryptomator/hub into feature/user-and-groups

overheadhunter committed:
eagerly fetch user details

overheadhunter committed:
adjust tests to new validations

overheadhunter committed:
eagerly fetch group data

overheadhunter committed:
fix vault count

overheadhunter committed:
correct indentation

[ci skip]

iammajid committed:
Fix type assertion for user roles

iammajid committed:
Fix removeUser

iammajid commented on frontend/src/components/authority/UserList.vue:
```diff
@@ -0,0 +1,258 @@
+        </svg>
+        <h3 class="mt-2 text-sm font-medium text-gray-900">{{ t('userList.filter.result.empty.title') }}</h3>
+        <p class="mt-1 text-sm text-gray-500">{{ t('userList.filter.result.empty.description') }}</p>
+      </div>
```
This empty state is intentional, there will always be at least one user, as the administrator viewing this page cannot delete themselves. Therefore, this state can never occur.

iammajid committed:
Fix empty state conditionals and error reset on retry

iammajid commented on frontend/src/i18n/en-US.json:
```diff
@@ -172,6 +195,54 @@
+  "groupList.create.button": "Create Group",
+  "groupList.group.created": "Created on",
+
+  "group.members.add":"Add Members",
```
Can be ignored, formatting is acceptable.

iammajid commented on frontend/src/common/backend.ts:
```diff
@@ -74,34 +76,71 @@ export type UserDto = {
   setupCode?: string;
 }
 
+export type UserDtoWithCounts = UserDto & {
+  groupsCount?: number;
+  devicesCount?: number;
+  accessibleVaults?: number;
+}
```
Already fixed, the field is named accessibleVaultCount, not accessibleVaults.

iammajid committed:
Use browser locale for sorting

iammajid committed:
Fix error ref initialization

overheadhunter committed:
fix realm roles

overheadhunter committed:
update effective_group_membership

overheadhunter committed:
Merge branch 'feature/user-and-groups' of https://github.com/cryptomator/hub into feature/user-and-groups

# Conflicts:
#	frontend/src/components/authority/UserEditCreate.vue

overheadhunter committed:
fix roles

overheadhunter committed:
show effective vault access in user details

overheadhunter committed:
1. group membership, 2. effective group membership

overheadhunter committed:
deduplicate

overheadhunter committed:
reload user details when changing group membership

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,357 @@
+	}
+
+	// TODO deduplicate with KeycloakAuthorityPuller
+	@Transactional
+	public User syncUser(String userId) {
+		UserResource userResource = realm.users().get(userId);
+		UserRepresentation keycloakUser = userResource.toRepresentation();
+
+		User dbUser = userRepo.findById(userId);
+		if (dbUser == null) {
+			dbUser = new User();
+			dbUser.setId(keycloakUser.getId());
+		}
+
+		dbUser.setName(keycloakUser.getUsername());
+		dbUser.setEmail(keycloakUser.getEmail());
+		dbUser.setFirstName(keycloakUser.getFirstName());
+		dbUser.setLastName(keycloakUser.getLastName());
+
+		var attrs = keycloakUser.getAttributes();
+		if (attrs != null && attrs.containsKey("picture")) {
+			var pictureAttr = attrs.get("picture");
+			dbUser.setPictureUrl(pictureAttr.isEmpty() ? null : pictureAttr.getFirst());
+		} else {
+			dbUser.setPictureUrl(null);
+		}
+
+		userRepo.persist(dbUser);
+		userRepo.flush();
+		return dbUser;
+	}
```
`updateUserRoles` is invoked by caller afterwards.

overheadhunter committed:
eagerly fetch stats in `GET /users/`

overheadhunter committed:
same json property name in frontend and backend

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/RealmRole.java:
```diff
@@ -0,0 +1,47 @@
+		throw new IllegalArgumentException("No matching RealmRole for Keycloak name: " + kcName);
+	}
+
+	public static Set<RealmRole> fromKcNames(Collection<String> kcNames) {
+		if (kcNames == null || kcNames.isEmpty()) {
+			return EnumSet.noneOf(RealmRole.class);
+		}
+		Set<RealmRole> roles = EnumSet.noneOf(RealmRole.class);
+		for (String kcName : kcNames) {
+			for (RealmRole role : values()) {
+				if (role.kcName.equals(kcName)) {
+					roles.add(role);
+				}
+			}
+		}
+		return roles;
+	}
```
by design

overheadhunter committed:
GroupEditCreate.vue: reduced `onSubmit` complexity

overheadhunter committed:
run formatter

overheadhunter committed:
fix NPE during user creation, if roles are empty

overheadhunter committed:
move route param evaluation to index.ts

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/StringArrayType.java:
```diff
@@ -36,31 +36,38 @@ public int hashCode(String[] x) {
 
 	@Override
-	public void nullSafeSet(PreparedStatement st, String[] value, int index, SharedSessionContractImplementor session) {
-		throw new UnsupportedOperationException("Read Only");
+	public void nullSafeSet(PreparedStatement st, String[] value, int index, SharedSessionContractImplementor session) throws SQLException {
+		session.doWork(connection -> {
+			var jdbcArray = connection.createArrayOf("VARCHAR", value == null ? new String[0] : value);
+			st.setArray(index, jdbcArray);
+		});
 	}
```
nope it does not implement AutoCloweable

overheadhunter replied:
I still disagree. The PreparedStatement still references the array and should be responsible for freeing it.

overheadhunter committed:
optimize `findByIdWithEagerDetails()`

avoid cartesian product explosion by only left-joining with tables that we don't expect to produce more than a handful of rows

overheadhunter committed:
optimize `findByIdWithEagerDetails()`

avoid cartesian product explosion

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/User.java:
```diff
@@ -164,6 +221,32 @@ public int hashCode() {
 	@ApplicationScoped
 	public static class Repository implements PanacheRepositoryBase<User, String> {
 
+		public PanacheQuery<User> findAllWithMetrics() {
+			return find("""
+					FROM User u
+					LEFT JOIN FETCH u.metrics m
+					""");
+		}
+
+		public User findByIdWithEagerDetails(String id) {
+			return find("""
+					FROM User u
+					LEFT JOIN FETCH u.directGroupMemberships dgm
+					LEFT JOIN FETCH u.accessibleVaults eva
+					LEFT JOIN FETCH eva.vault v
+					LEFT JOIN FETCH u.devices d
+					LEFT JOIN FETCH u.legacyDevices ld
+					WHERE u.id = :id
+					""", Parameters.with("id", id)).singleResultOptional().orElse(null);
+		}
```
fixed in 85940088540ab72ac2d0fb93ee6ea6ab0a3c13b3

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/Group.java:
```diff
@@ -40,6 +47,15 @@ public int getMemberSize() {
 	@ApplicationScoped
 	public static class Repository implements PanacheRepositoryBase<Group, String> {
 
+		public Group findByIdWithEagerDetails(String id) {
+			return find("""
+					FROM Group g
+					LEFT JOIN FETCH g.members
+					LEFT JOIN FETCH g.accessibleVaults
+					WHERE g.id = :id
+					""", Parameters.with("id", id)).singleResultOptional().orElse(null);
+		}
```
fixed in 82d94215226942b1f867a73cf793830cd46efc79

overheadhunter committed:
cleanup

make sonarcloud happy

overheadhunter committed:
remove unused import

[ci skip]

iammajid committed:
Fix legacy device list info prop

iammajid commented on frontend/src/components/authority/UserEditCreate.vue:
```diff
@@ -0,0 +1,509 @@
+        errors.value.username = t('userEditCreate.error.userAlreadyExists');
+      }
+    } else {
+      submitError.value = error instanceof Error ? error.message : 'An error occurred';
```
This is consistent with the rest of the codebase, all error fallback messages use hardcoded strings.

iammajid commented on frontend/src/components/authority/GroupDetail.vue:
```diff
@@ -0,0 +1,134 @@
+  vaults: []
+});
+
+function onMembersSaved(newMembers: AuthorityDto[]) {
+  const ids = new Set(group.value.members.map(u => u.id));
+  newMembers.forEach(u => { if (!ids.has(u.id)) group.value.members.push(u); });
+  group.value.members.sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }));
+}
```
Both UserDto and GroupDto define name: string (not optional), so name will never be undefined here.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakRealmRoles.java:
```diff
@@ -0,0 +1,29 @@
+	@ConfigProperty(name = "hub.keycloak.realm")
+	String keycloakRealm;
+
+	private final EnumMap<RealmRole, RoleRepresentation> cachedRoles = new EnumMap<>(RealmRole.class);
+
+	public RoleRepresentation getRealmRole(RealmRole realmRole) {
+		return cachedRoles.computeIfAbsent(realmRole, this::fetchRealmRole);
+	}
```
since the mapping should be idempotent, concurrent re-computations should result in the same value. So race conditions are merely an academic concern in this case.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,357 @@
+
+	public GroupRepresentation updateGroup(String groupId, String name, String pictureUrl) {
+		GroupResource groupResource = realm.groups().group(groupId);
+		GroupRepresentation group = groupResource.toRepresentation();
+
+		if (!name.isBlank()) {
+			group.setName(name);
+		}
+
+		if (pictureUrl != null) {
+			var attrs = setPicture(group.getAttributes(), pictureUrl);
+			group.setAttributes(attrs);
+		}
+
+		groupResource.update(group);
+		syncGroup(groupId);
+		return groupResource.toRepresentation();
+	}
```
caller already verifies name is non-null. we might want to introduce jspecify annotations, though. Please create a corresponding issue.

overheadhunter committed:
don't define `@JoinTable` on both sides

iammajid committed:
Merge branch 'feature/user-and-groups' of github.com:cryptomator/hub into feature/user-and-groups

iammajid committed:
Fix accessibility and error handling issues

iammajid committed:
Use button elements for remove actions

iammajid commented on frontend/src/components/authority/UserDetail.vue:
```diff
@@ -0,0 +1,142 @@
+        <UserInfo :user="user"/>
+        <!-- Devices -->
+        <UserDeviceList :devices="user.devices" :page-size="10" :title="t('user.detail.devices')"/>      
+        <UserDeviceList :devices="user.legacyDevices" :page-size="10" :visible="user.legacyDevices.length != 0" :title="t('legacyDeviceList.title')" :info="t('legacyDeviceList.title')"/>
```
This has already been fixed.

iammajid commented on frontend/src/components/authority/GroupEditCreate.vue:
```diff
@@ -0,0 +1,244 @@
+      pictureUrl.value = group.pictureUrl ?? '';
+      initialGroupData.value = {
+        id: group.id,
+        name: group.name,
+        pictureUrl: group.pictureUrl ?? ''
+      };
+    } catch (error) {
+      console.error('Failed to fetch group:', error);
+    }
```
This is consistent with the current codebase pattern

iammajid committed:
Remove redundant pictureUrl fallback in GroupAddMemberDialog

iammajid committed:
Fill in missing group pictures in user detail view

overheadhunter committed:
fix signature (null-safe methods)

overheadhunter committed:
consolidate DTO types

overheadhunter committed:
Merge branch 'develop' into feature/user-and-groups

overheadhunter committed:
update CHANGELOG

[ci skip]

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,431 @@
+	}
+
+	public boolean isUserReadOnly(String userId) {
+		RealmResource realm = keycloak.realm(keycloakRealm);
+		try {
+			UserResource userResource = realm.users().get(userId);
+			var federatedIdentities = userResource.getFederatedIdentity();
+			return !federatedIdentities.isEmpty();
+		} catch (Exception e) {
+			LOG.warn("Failed to check federated identity for user {}", userId, e);
+			return false;
+		}
+	}
```
@SailReal I agree with this finding. How shall we proceed? Let the exception bubble up or default to `true`?

SailReal replied:
@overheadhunter Bubble up the exception because the UI should behave differently if a user should be indeed treated as read only VS there was a network hiccup.

overheadhunter committed:
edit user: allow changing email

overheadhunter committed:
fix group count in user list

overheadhunter committed:
fix test

iammajid committed:
Fix password update in user edit form

iammajid committed:
Deduplicate vaults in user detail view

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/User.java:
```diff
@@ -126,6 +171,23 @@ public void setDevices(Set<Device> devices) {
+			joinColumns = @JoinColumn(name = "member_id", referencedColumnName = "id"),
+			inverseJoinColumns = @JoinColumn(name = "group_id", referencedColumnName = "id")
+	)
+	public Set<Group> directGroupMemberships = new HashSet<>();
```
fixed in 3fab527

overheadhunter commented on frontend/src/components/authority/GroupAddMemberDialog.vue:
```diff
@@ -0,0 +1,142 @@
+
+  try {
+    // Add each member to the group via API
+    for (const member of newMembers.value) {
+      await backend.groups.addMember(props.groupId, member.id);
+    }
```
we will keep sequential processing for now, as the keycloak api doesn't allow bulk-adding of group members.

overheadhunter commented on frontend/src/common/formvalidator.ts:
```diff
@@ -0,0 +1,130 @@
+  /**
+   * Validates user form data
+   */
+  static validateUser(data: {
+    firstName?: string,
+    lastName?: string,
+    username: string,
+    email: string,
+    password: string,
+    passwordConfirm: string,
+    isEditMode: boolean,
+    pictureUrl?: string | null,
+    isValidImageUrl?: boolean
+  }): ValidationResult {
+    const errors: Record<string, string> = {};
+
+    // Required fields validation
+    if (!data.firstName?.trim()) errors.firstName = t('userEditCreate.validation.required');
+    if (!data.lastName?.trim()) errors.lastName = t('userEditCreate.validation.required');
+    if (!data.username.trim()) errors.username = t('userEditCreate.validation.required');
```
the type reflects the fact that legacy data may still have missing first and last names. however, newly added data should require both.

overheadhunter commented on frontend/src/components/authority/GroupEditCreate.vue:
```diff
@@ -0,0 +1,225 @@
+    // Redirect after successful save
+    setTimeout(() => {
+      router.push('/app/groups');
+    }, 1000);
```
@iammajid yesterday we discussed that group and user redirects should be treated the same anyway. What do you think about redirecting without a delay? We need to decide whether we want to go to group details or back to the list, though.

iammajid replied:
@overheadhunter Yes, removing the delay makes sense. The redirect targets are already correct, create goes to list, edit goes to detail page. This feels natural because you return to where you came from. Just need to remove the setTimeout. Or how do you see it?

overheadhunter replied:
agreed. Can you do this?

overheadhunter replied:
the `debouncedGroupSaved` stuff can be removed as well, can't it?

overheadhunter replied:
removed debounce in 050e2e1

overheadhunter committed:
cleanup, fix tests

overheadhunter committed:
fail hard, if we can't determine federation status

as discussed in https://github.com/cryptomator/hub/pull/376/changes#r2670837273

overheadhunter committed:
refactor callback to use `defineEmits`

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -290,4 +302,158 @@ public static TrustedUserDto fromEntity(EffectiveWot entity) {
+			return Response.created(URI.create("./" + user.getId()))
+					.entity(UserDto.justPublicInfo(user))
+					.build();
+		} catch (ClientErrorException e) {
+			// Return 409 with specific error message (EMAIL_EXISTS or USERNAME_EXISTS)
+			return Response.status(Response.Status.CONFLICT)
+					.entity(e.getMessage())
+					.type(MediaType.TEXT_PLAIN)
+					.build();
+		}
```
acceptable in this case

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+					var location = response.getHeaderString("Location");
+					yield location.substring(location.lastIndexOf('/') + 1);
+				}
+				case 409 -> {
+					String body = response.readEntity(String.class);
+					String errorMessage;
+					if (body != null && body.contains("same email")) {
+						errorMessage = "EMAIL_EXISTS";
+					} else if (body != null && body.contains("same username")) {
+						errorMessage = "USERNAME_EXISTS";
+					} else {
+						errorMessage = "User already exists";
+					}
+					throw new ClientErrorException(errorMessage, Response.Status.CONFLICT);
```
acceptable in this case

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+
+		// groups need to be set after creation, see https://github.com/keycloak/keycloak/discussions/8552
+		if (groupIds != null && !groupIds.isEmpty()) {
+			for (String groupId : groupIds) {
+				try {
+					userResource.joinGroup(groupId);
+				} catch (WebApplicationException e) {
+					LOG.warn("Failed to add user {} to group {}", userId, groupId, e);
+					// TODO: shall we fail the whole user creation here? undo previous steps?
+				}
+			}
+			effectiveGroupMembershipRepo.updateGroups(groupIds);
+		}
```
changes can't be made in atomic transaction due to limitations of upstream API. In case of error, the user needs to manually fix data inconsistencies.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+		var dbMembers = userRepo.findByIds(keycloakMembers).toList();
+		dbGroup.getMembers().clear();
+		dbGroup.getMembers().addAll(dbMembers);
+		groupRepo.persist(dbGroup);
+		effectiveGroupMembershipRepo.updateGroups(Set.of(dbGroup.getId()));
+		return dbGroup;
+	}
+
+	@Transactional
+	public void addUserToGroup(String groupId, String userId) {
+		realm.groups().group(groupId).toRepresentation(); // trigger NotFoundException if group does not exist
+		UserResource userResource = realm.users().get(userId);
+		userResource.joinGroup(groupId);
+
+		syncGroup(groupId);
+	}
```
@iammajid do you know whether `userResource.joinGroup(groupId)` will trigger 404 for unknown groups?

[the api docs](https://www.keycloak.org/docs-api/latest/rest-api/index.html#_put_adminrealmsrealmusersuser_idgroupsgroupid) do mention `404`, however it is unclear if this applies to all three path parameters in `PUT /admin/realms/{realm}/users/{user-id}/groups/{groupId}`...

iammajid replied:
@overheadhunter Tested: `PUT /admin/realms/.../users/{userId}/groups/{fakeGroupId}` returns 404. So `joinGroup()` handles unknown groups itself the `toRepresentation()` call is redundant. Should I remove it?

overheadhunter replied:
yes, I'll take care of it

overheadhunter replied:
fixed in 7bc56f0 and dbbeace

iammajid committed:
Remove redirect delay in GroupEditCreate

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+		}
+	}
+
+	// Group management methods
+
+	public GroupRepresentation createGroup(String name, String pictureUrl) {
+		GroupRepresentation group = new GroupRepresentation();
+		group.setName(name);
+
+		if (pictureUrl != null && !pictureUrl.isBlank()) {
+			group.setAttributes(Map.of("picture", List.of(pictureUrl)));
+		}
+
+		final String groupId;
+		try (var response = realm.groups().add(group)) {
+			groupId = switch (response.getStatus()) {
+				case 201 -> {
+					var location = response.getHeaderString("Location");
+					yield location.substring(location.lastIndexOf('/') + 1);
+				}
+				case 409 -> throw new ClientErrorException("GROUP_NAME_EXISTS", Response.Status.CONFLICT);
+				default -> throw new InternalServerErrorException("Failed to create group in Keycloak. Status: " + response.getStatus());
+			};
+		}
+
+		syncGroup(groupId);
+		return realm.groups().group(groupId).toRepresentation();
+	}
```
Making this method `@Transactional` doesn't help, as nothing can be rolled back here. All the changes happen outside of the transaction via ReST calls.

Data inconsistencies are tolerable, as Keycloak is the single source of truth. DB is always just a read-only replica. Inconsistencies will solve themselves eventually on later syncs.

overheadhunter committed:
allow rollbacks by first syncing to DB, then to KC

overheadhunter commented on frontend/src/components/authority/UserGroupsList.vue:
```diff
@@ -0,0 +1,159 @@
+const { t } = useI18n({ useScope: 'global' });
+
+const props = defineProps<{
+  user: UserDtoWithDetails;
+  userId: string;
+  groups: GroupDto[];
+  pageSize: number;
+}>();
```
true. @iammajid is this even supposed to be configured by the surrounding component via a param?

iammajid replied:
@overheadhunter No, i think it's always 10 everywhere. We can remove the prop and hardcode it.

overheadhunter replied:
fixed in 4b9a22b

overheadhunter committed:
remove inefficient `toRepresentation()` web call

overheadhunter committed:
allow rollbacks by first syncing to DB, then to KC

also improved query efficiency

overheadhunter committed:
made PAGE_SIZE a constant

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+
+		syncGroup(groupId);
+	}
+
+	@Transactional
+	public void removeUserFromGroup(String groupId, String userId) {
+		realm.groups().group(groupId).toRepresentation(); // trigger NotFoundException if group does not exist
+		UserResource userResource = realm.users().get(userId);
+		userResource.leaveGroup(groupId);
+
+		syncGroup(groupId);
+	}
```
fixed in 7bc56f0 and dbbeace

overheadhunter committed:
accessibility

overheadhunter committed:
cleanup, avoid dev confusion about new methods

overheadhunter committed:
fix handling of picture url in user/group form

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/Group.java:
```diff
@@ -40,9 +57,40 @@ public int getMemberSize() {
+			}
+			// 2. fetch accessible vaults separately to avoid cartesian product explosion:
+			Hibernate.initialize(group.accessibleVaults);
+			return group;
+		}
+
 		public long deleteByIds(Collection<String> ids) {
 			return Batch.of(200).run(ids, 0L, (batch, result) -> result + delete("id IN :ids", Parameters.with("ids", batch)));
 		}
 
+		public void addMember(String groupId, String memberId) {
+			getEntityManager()
+					.createNamedQuery("Group.addMember")
+					.setParameter("groupId", groupId)
+					.setParameter("memberId", memberId)
+					.executeUpdate();
+		}
+
+		public void removeMember(String groupId, String memberId) {
+			getEntityManager()
+					.createNamedQuery("Group.removeMember")
+					.setParameter("groupId", groupId)
+					.setParameter("memberId", memberId)
+					.executeUpdate();
+		}
```
this is acceptable, I've added a warning in javadoc

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,351 @@
+		return dbUser;
+	}
+
+	// TODO deduplicate with KeycloakAuthorityPuller
+	@Transactional
+	public Group syncGroup(String groupId) {
+		GroupResource groupResource = realm.groups().group(groupId);
+		GroupRepresentation keycloakGroup = groupResource.toRepresentation();
+
+		Group dbGroup = groupRepo.findById(groupId);
+		if (dbGroup == null) {
+			dbGroup = new Group();
+			dbGroup.setId(keycloakGroup.getId());
+		}
+
+		dbGroup.setName(keycloakGroup.getName());
+
+		var attrs = keycloakGroup.getAttributes();
+		if (attrs != null && attrs.containsKey("picture")) {
+			var pictureAttr = attrs.get("picture");
+			dbGroup.setPictureUrl(pictureAttr.isEmpty() ? null : pictureAttr.getFirst());
+		} else {
+			dbGroup.setPictureUrl(null);
+		}
+
+		// Sync members
+		var keycloakMembers = groupResource.members().stream().map(UserRepresentation::getId).toList();
+		var dbMembers = userRepo.findByIds(keycloakMembers).toList();
+		dbGroup.getMembers().clear();
+		dbGroup.getMembers().addAll(dbMembers);
+		groupRepo.persist(dbGroup);
+		effectiveGroupMembershipRepo.updateGroups(Set.of(dbGroup.getId()));
+		return dbGroup;
+	}
```
can you please create an issue for the dedup TODOs on syncUser and syncGroup?

overheadhunter committed:
no need to recompute jdenticon dynamically

[ci skip]

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/entities/Group.java:
```diff
@@ -3,31 +3,48 @@
 	)
 	private Set<Authority> members = new HashSet<>();
 
+	@Immutable
+	@OneToMany(mappedBy = "authority", fetch = FetchType.LAZY)
+	public Set<VaultAccess> accessibleVaults = new HashSet<>();
```
acceptable for immutable properties

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,366 @@
+
+	public GroupRepresentation updateGroup(String groupId, String name, String pictureUrl) {
+		GroupResource groupResource = realm.groups().group(groupId);
+		GroupRepresentation group = groupResource.toRepresentation();
+
+		if (!name.isBlank()) {
+			group.setName(name);
+		}
+
+		if (pictureUrl != null) {
+			var attrs = setPicture(group.getAttributes(), pictureUrl);
+			group.setAttributes(attrs);
+		}
+
+		groupResource.update(group);
+		syncGroup(groupId);
+		return groupResource.toRepresentation();
+	}
```
name is protected by `jakarta.validation.constraints.NotNull`

overheadhunter committed:
reorder methods to ensure consistency

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,366 @@
+
+		// groups need to be set after creation, see https://github.com/keycloak/keycloak/discussions/8552
+		if (groupIds != null && !groupIds.isEmpty()) {
+			for (String groupId : groupIds) {
+				try {
+					userResource.joinGroup(groupId);
+				} catch (WebApplicationException e) {
+					LOG.warn("Failed to add user {} to group {}", userId, groupId, e);
+					// TODO: shall we fail the whole user creation here? undo previous steps?
+				}
+			}
+			effectiveGroupMembershipRepo.updateGroups(groupIds);
+		}
```
As already mentioned in other review comments, the upstream API doesn't allow atomic mutation of multiple data. The good news: `effectiveGroupMembershipRepo.updateGroups(groupIds)` is idempotent and is designed for eventually consistent data.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -178,7 +188,7 @@ public UserDto getMe(@QueryParam("withDevices") boolean withDevices, @QueryParam
 			deviceDtos = Set.of();
 		}
-		return new UserDto(user.getId(), user.getName(), user.getPictureUrl(), user.getEmail(), user.getLanguage(), deviceDtos, user.getEcdhPublicKey(), user.getEcdsaPublicKey(), user.getPrivateKeys(), user.getSetupCode());
+		return new UserDto(user.getId(), user.getName(), user.getPictureUrl(), user.getEmail(), user.getFirstName(), user.getLastName(), user.getLanguage(), Set.of(user.getRealmRoles()), deviceDtos, user.getEcdhPublicKey(), user.getEcdsaPublicKey(), user.getPrivateKeys(), user.getSetupCode());
```
No, `Set.of(E...)` will create a `Set<String>`.

See [JavaDoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Set.html#of(E...)):

> This method also accepts a single array as an argument. The element type of the resulting set will be the component type of the array, and the size of the set will be equal to the length of the array.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/api/UsersResource.java:
```diff
@@ -196,13 +206,13 @@ public UserDto getMe(@QueryParam("withDevices") boolean withDevices, @QueryParam
 			return DeviceResource.DeviceDto.fromEntity(d, event);
 		}).collect(Collectors.toSet());
-		return new UserDto(user.getId(), user.getName(), user.getPictureUrl(), user.getEmail(), user.getLanguage(), deviceDtos, user.getEcdhPublicKey(), user.getEcdsaPublicKey(), user.getPrivateKeys(), user.getSetupCode());
+		return new UserDto(user.getId(), user.getName(), user.getPictureUrl(), user.getEmail(), user.getFirstName(), user.getLastName(), user.getLanguage(), Set.of(user.getRealmRoles()), deviceDtos, user.getEcdhPublicKey(), user.getEcdsaPublicKey(), user.getPrivateKeys(), user.getSetupCode());
```
No, `Set.of(E...)` will create a `Set<String>`.

See [JavaDoc](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Set.html#of(E...)):

> This method also accepts a single array as an argument. The element type of the resulting set will be the component type of the array, and the size of the set will be equal to the length of the array.

overheadhunter committed:
fix 4d1dd96

overheadhunter committed:
fail early

overheadhunter committed:
same submit handling for users and groups

remove debounce stuff
[ci skip]

overheadhunter committed:
consider user data changed for non-empty password

[ci skip]

overheadhunter commented on frontend/src/components/authority/UserEditCreate.vue:
```diff
@@ -0,0 +1,436 @@
+                  <button type="button" class="inline-flex justify-center py-2 px-4 border border-gray-300 shadow-sm text-sm font-medium rounded-md bg-white text-gray-700 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary" @click="cancelAction">
+                    {{ t('common.cancel') }}
+                  </button>
+                  <button type="submit" :disabled="processing || !userDataHasUnsavedChanges || password !== passwordConfirm" class="inline-flex justify-center py-2 px-4 border border-transparent shadow-sm text-sm font-medium rounded-md text-white bg-primary hover:bg-primary-d1 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary disabled:opacity-50 disabled:hover:bg-primary disa ...
```
fixed in 871d5b0

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		return userResource.toRepresentation();
+	}
+
+	@Transactional
+	public void deleteUser(String userId) {
+		if (isUserReadOnly(userId)) {
+			throw new ForbiddenException("User has a federated identity and cannot be deleted");
+		}
+
+		// 1. delete from db (roll back if kc deletion fails):
+		userRepo.deleteById(userId);
+
+		// 2. delete from kc:
+		try (var response = realm.users().delete(userId)){
+			if (response.getStatus() != 204) {
+				throw new InternalServerErrorException("Failed to delete user in Keycloak. Status: " + response.getStatus());
+			}
+		}
+	}
```
if kc deletion failed, db will rollback, both systems are in sync. the reverse order is only necessary, where no strong consistency can be achieved.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		return dbGroup;
+	}
+
+	@Transactional
+	public void addUserToGroup(String groupId, String userId) {
+		// 1. sync to db (roll back if kc update fails):
+		try {
+			groupRepo.addMember(groupId, userId);
+			effectiveGroupMembershipRepo.updateGroups(List.of(groupId));
+		} catch (PersistenceException e) { // caused by foreign key constraint violation
+			throw new NotFoundException("Failed to add member " + userId + " to group " + groupId);
+		}
+
+		// 2. sync to kc:
+		realm.users().get(userId).joinGroup(groupId);
+	}
```
if kc deletion failed, db will rollback, both systems are in sync. the reverse order is only necessary, where no strong consistency can be achieved.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		realm.users().get(userId).joinGroup(groupId);
+	}
+
+	@Transactional
+	public void removeUserFromGroup(String groupId, String userId) {
+		// 1. sync to db (roll back if kc update fails):
+		groupRepo.removeMember(groupId, userId);
+		effectiveGroupMembershipRepo.updateGroups(List.of(groupId));
+
+		// 2. sync to kc:
+		realm.users().get(userId).leaveGroup(groupId);
+	}
```
if kc deletion failed, db will rollback, both systems are in sync. the reverse order is only necessary, where no strong consistency can be achieved.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		realm.users().get(userId).leaveGroup(groupId);
+	}
+
+	@Transactional
+	public void updateUserRoles(String userId, Set<RealmRole> roles) {
+		// remove roles that are not in the provided set:
+		var rolesToRemove = EnumSet.allOf(RealmRole.class);
+		rolesToRemove.removeAll(roles);
+
+		// set roles that are in the provided set:
+		var rolesToSet = EnumSet.noneOf(RealmRole.class);
+		rolesToSet.addAll(roles);
+
+		// 1. sync to db (roll back if kc update fails):
+		User dbUser = userRepo.findByIdOptional(userId).orElseThrow(NotFoundException::new);
+		dbUser.setRealmRoles(rolesToSet.stream().map(RealmRole::kcName).toArray(String[]::new));
+		userRepo.persist(dbUser);
+
+		// 2. sync to kc:
+		UserResource userResource = realm.users().get(userId);
+		var roleMappings = userResource.roles().realmLevel();
+		if (!rolesToRemove.isEmpty()) {
+			roleMappings.remove(rolesToRemove.stream().map(realmRoles::getRealmRole).toList());
+		}
+		if (!rolesToSet.isEmpty()) {
+			roleMappings.add(rolesToSet.stream().map(realmRoles::getRealmRole).toList());
+		}
+	}
```
if kc deletion failed, db will rollback, both systems are in sync. the reverse order is only necessary, where no strong consistency can be achieved.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		return groupResource.toRepresentation();
+	}
+
+	@Transactional
+	public void deleteGroup(String groupId) {
+		// 1. delete from db (roll back if kc deletion fails):
+		groupRepo.deleteById(groupId);
+
+		// 2. delete from kc:
+		realm.groups().group(groupId).remove();
+	}
```
if kc deletion failed, db will rollback, both systems are in sync. the reverse order is only necessary, where no strong consistency can be achieved.

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/keycloak/KeycloakAdminService.java:
```diff
@@ -0,0 +1,364 @@
+		GroupResource groupResource = realm.groups().group(groupId);
+		GroupRepresentation group = groupResource.toRepresentation();
+
+		if (!name.isBlank()) {
+			group.setName(name);
+		}
```
we have already established the fact that name is null-protected by jakarta validations.

overheadhunter commented on frontend/src/components/authority/GroupEditCreate.vue:
```diff
@@ -0,0 +1,211 @@
+  resetGroupData();
+});
+
+function removePicture() {
+  data.pictureUrl = undefined;
+}
```
no, GroupDto defines `pictureUrl?: string`

overheadhunter committed:
made batch processing null-safe

fixes NPE during user creation

overheadhunter committed:
debounce image url validation

overheadhunter committed:
show user-facing error

overheadhunter committed:
address minor issues from code review

overheadhunter committed:
avoid outdated data during form validation

debounce makes `isValidImageUrl` eventually consistent

overheadhunter committed:
accessibility fixes

overheadhunter committed:
consistent initial data

overheadhunter committed:
add timeout to attempted image loading

iammajid committed:
Fix roles placeholder visibility

overheadhunter commented on frontend/src/components/authority/UserDeviceList.vue:
```diff
@@ -0,0 +1,224 @@
+      </table>
+    </div>
+
+    <!-- Pagination -->
+    <div v-if="showPaginationDevice" class="bg-gray-50 border-t border-gray-200">
+      <nav class="flex items-center justify-between px-4 py-3 sm:px-6" :aria-label="t('common.pagination')">
+        <div class="hidden sm:block">
+          <i18n-t keypath="auditLog.pagination.showing" scope="global" tag="p" class="text-sm text-gray-700">
+            <span class="font-medium">{{ paginationBeginDevice }}</span>
+            <span class="font-medium">{{ paginationEndDevice }}</span>
+          </i18n-t>
+        </div>
```
component just used once.

overheadhunter commented on frontend/src/components/authority/UserDeviceList.vue:
```diff
@@ -0,0 +1,224 @@
+// DEVICES – search & pagination
+// ---------------------------------------------------------------------------
+
+// Filter und Pagination
+const deviceQuery = ref('');
+const filteredDevices = computed(() =>
+  props.devices.filter((device) =>
+    device.name.toLowerCase().includes(deviceQuery.value.toLowerCase())
+  )
+);
```
name is non-null

overheadhunter commented on frontend/src/components/authority/UserEditCreate.vue:
```diff
@@ -0,0 +1,425 @@
+                  <button type="button" class="inline-flex justify-center py-2 px-4 border border-gray-300 shadow-sm text-sm font-medium rounded-md bg-white text-gray-700 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary" @click="cancelAction">
+                    {{ t('common.cancel') }}
+                  </button>
+                  <button type="submit" :disabled="processing || !userDataHasUnsavedChanges || password !== passwordConfirm" class="inline-flex justify-center py-2 px-4 border border-transparent shadow-sm text-sm font-medium rounded-md text-white bg-primary hover:bg-primary-d1 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary disabled:opacity-50 disabled:hover:bg-primary disa ...
+                    <span>{{ props.mode === 'EDIT' ? t('common.save') : t('common.create') }}</span>
+                  </button>
```
nope, `outline-none` and `outline-hidden` are not the same!

overheadhunter committed:
unique element ids when including component twice

overheadhunter committed:
reset password fields

overheadhunter commented on frontend/src/components/authority/UserDeviceList.vue:
```diff
@@ -0,0 +1,224 @@
+            </tr>
+          </template>
+
+          <tr v-if="!filteredDevices.length">
+            <td colspan="5" class="py-4 px-6 text-sm text-gray-500 text-center">
+              {{ t(deviceQuery ? 'common.nothingFound' : 'common.none') }}
+            </td>
+          </tr>
```
this is more complicated than I thought. the number of cols depends on whether a mobile or desktop page is rendered. please create a separate issue to fix the table layouts

overheadhunter committed:
parameterized info text

overheadhunter commented on frontend/src/components/authority/UserDeviceList.vue:
```diff
@@ -0,0 +1,225 @@
+                </div>
+              </td>
+              <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
+                {{ new Date(device.creationTime).toISOString().slice(0, 16).replace('T', ' ') }}
```
please create an issue for this

overheadhunter committed:
more accessibility and css fixes

overheadhunter commented on frontend/src/components/authority/UserDeviceList.vue:
```diff
@@ -0,0 +1,225 @@
+            <tr class="hidden sm:table-row">
+              <td class="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900">
+                <div class="flex items-center gap-2">
+                  <span v-if="device.type == 'BROWSER'" :title="'Browser'">
+                    <WindowIcon class="h-5 w-5 text-gray-500" aria-hidden="true" />
+                  </span>
+                  <span v-else-if="device.type == 'DESKTOP'" :title="'Desktop'">
+                    <ComputerDesktopIcon class="h-5 w-5 text-gray-500" aria-hidden="true" />
+                  </span>
+                  <span v-else-if="device.type == 'MOBILE'" :title="'Mobile'">
+                    <DevicePhoneMobileIcon class="h-5 w-5 text-gray-500" aria-hidden="true" />
+                  </span>
```
create an issue for this

overheadhunter commented on frontend/src/components/authority/GroupMemberList.vue:
```diff
@@ -0,0 +1,165 @@
+              <button type="button" class="cursor-pointer text-red-600 hover:text-red-900" :title="t('common.leave')" @click="showDeleteDialog(member)">{{ t('common.remove') }}</button>
+            </td>
+          </tr>
+          <tr v-if="!filteredUsers.length">
+            <td colspan="2" class="py-4 px-4 text-sm text-center text-gray-500">
+              {{ t(userQuery ? 'group.members.search.empty' : 'common.none') }}
+            </td>
+          </tr>
```
since the text is either `group.members.search.empty` or `common.none`, I think it is correct to also show the message when the filtered results are empty.
